sales/src/execute.rs

```diff
@@ -0,0 +1,209 @@
+use std::{cell::RefCell, rc::Rc};
+
+use burnt_glue::response::Response;
+use cosmwasm_std::{
+    BankMsg, Coin, CosmosMsg, CustomMsg, Deps, DepsMut, Env, MessageInfo, Timestamp, Uint64,
+};
+use cw721_base::{state::TokenInfo, MintMsg};
+use cw_storage_plus::Item;
+use ownable::Ownable;
+use sellable::Sellable;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{errors::ContractError, state::PrimarySale, Sales};
+
+impl<'a, T, C, E, Q> Sales<'a, T, C, E, Q>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    Q: CustomMsg,
+    E: CustomMsg,
+    C: CustomMsg,
+{
+    pub fn new(
+        sellable_module: Rc<RefCell<Sellable<'a, T, C, E, Q>>>,
+        primary_sales: Item<'a, Vec<PrimarySale>>,
+    ) -> Self {
+        Self {
+            sellable: sellable_module,
+            primary_sales,
+        }
+    }
+
+    pub fn add_primary_sales(
+        &mut self,
+        total_supply: Uint64,
+        start_time: Uint64,
+        end_time: Uint64,
+        price: Vec<Coin>,
+        deps: &mut DepsMut,
+        env: Env,
+        info: &MessageInfo,
+    ) -> Result<Response, ContractError> {
+        let ownable = &self.sellable.borrow().ownable;
+        check_ownable(&deps.as_ref(), &env, &info, &ownable.borrow())?;
+        // make sure no active primary sale
+        let mut primary_sales = self.primary_sales.load(deps.storage).unwrap();
+        for sale in &primary_sales {
+            if !sale.disabled || sale.end_time.gt(&env.block.time) {
+                return Err(ContractError::OngoingPrimarySaleError);
+            }
+        }
+        let primary_sale = PrimarySale {
```

Comment:
After creating the new struct I detailed above, you can simply implement a `To` or `From` instance to convert it into a primary sale object (setting the default values like you do here). That will reduce this to `let primary_sale = sale_spec.into()` where `sale_spec` is the argument passed in to this function.

Reply:
Nice and elegant